### ISSUE TITLE
confused about the teacher and student training

### ISSUE THREAD

**Issue by ccc013** [jlezama/disentangling-jacobian#1]
Hi, thanks for sharing the codes! And it's a good job!

But I have confused about the teacher and student model training because I  see you define the student and teacher model at the same time on the train.py:

```
# build the Student model
logger.info('build Student model')
ae = AutoEncoder(params).cuda()

# build the Teacher model
params2 = params
params2.max_fm = 512 + params.n_attr
logger.info('build Teacher model')
ae_teacher = AutoEncoder(params2).cuda()
```

But in the paper, it says we should firstly train the teacher autoencoder, and create the student model after completing the training the teacher model.

So could you add a description about how to use the codes in conditional_image_generation?

Thanks!

**Comment by jlezama** [comment]
Thanks for pointing this out. I fixed these comments and added some clarification.

The training scheme (which is inside src/training.py) depends on the value of the parameter lambda_jacobian. 

- If lambda_jacobian is 0, then _ae_ will be a standard trained autoencoder and _ae_teacher_ will not be used. 
- When params.lambda_jacobian >0, _ae_ will be the student and _ae_teacher_ will be the loaded teacher model from where to distill the Jacobian (the path to this model goes into  params.ae_teacher_reload).

**Comment by ccc013** [comment]
@jlezama Thanks for replying my question and updating the codes, README.md, and train_simple.sh, now I know how to use your code to train the teacher and student models.

**Comment by ccc013** [comment]
@jlezama I'm sorry for bothering you again.

After updating the codes, and I also successfully run and train the model, and I have two questions:

1. Could you tell me **how much time you use to train the teacher and student model totally**? Because I found it takes 17 hours to train the teacher model only in phase A that you say in the train_examples.sh, and I use one P40 Nvidia Graphics card, and I also try to use four cards, but no obvious speed up the training.

2. In addition, I also download the pre-trained model and run the web_server.py, I found if I set a large value for some attributes, the face in the Manipulation result would be distortion obviously, but I see the demo video you upload, no such situation has occurred. By the way, I use python3, not python2.

some bad examples like:

![image](https://user-images.githubusercontent.com/7010876/58525440-4fae2580-81fe-11e9-95b3-ba150470682f.png)

![image](https://user-images.githubusercontent.com/7010876/58525582-dd8a1080-81fe-11e9-87a3-d6be69ba0a04.png)


At last, I'm sorry for my poor English and looking forward to your reply.

Thanks!

**Comment by jlezama** [comment]
Hello again and thanks for trying it out. Indeed each attribute has a different working range that depends on the values the decoder has seen during training, and on how much it can extrapolate. 

However there is definitely something wrong in your result. I tried the same values and it gives a reasonable result (see below). Also the predicted numbers are a bit different for me. I'm using Python 2.7 and Pytorch 0.4.1. Just in case, make sure the model is in eval mode.

<img width="1440" alt="1st_example" src="https://user-images.githubusercontent.com/3524429/58557746-e6c6ad80-81fd-11e9-8c64-519e895dd81e.png">

Regarding training time, yes, training this model takes multiple days.

**Comment by ccc013** [comment]
Thanks for replying to me so quickly!

I'm sure the model is in eval mode, and I will try to run the web_server.py in Python2.7 and Pytorch 0.4.1 to see the result.

By the way, I see the Original image in your result is not the same as mine, so would the problem caused by the input image? My celebA dataset images are 178\*218 size, but I think your image size is 178\*178.

**Comment by jlezama** [comment]
that might be the reason... there is a pre-processing step included in data/preprocess.py
